# **mini-Hackathon 1: สถิติหน่วยจำหน่ายไฟฟ้าแยกตามเขต ⚡⚡⚡**
**รายละเอียดชุดข้อมูล(Dataset Description)**: ชุดข้อมูลนี้แสดง สถิติหน่วยจำหน่ายไฟฟ้า (Electricity Consumption) ของพื้นที่ให้บริการของการไฟฟ้านครหลวง (MEA) โดยจำแนกตาม เขตการจำหน่ายไฟฟ้า และช่วงเวลารายเดือน

ข้อมูลครอบคลุมช่วงปี 2562 – 2568 แสดงสถิติจำนวนหน่วยจำหน่าย (ล้านหน่วย (GWh)) แยกตามการไฟฟ้านครหลวงเขตการใช้ไฟฟ้า

**แหล่งที่มาข้อมูล**: https://data.go.th/dataset/sales-unit-statistics-by-area

**ชุดข้อมูลใน THackle**: https://www.thackle.or.th/th/dataset/97

---

**ข้อมูลเพิ่มเติม**
การไฟฟ้านครหลวง มีทั้งหมด 18 เขตทำการ ดังนี้
1. การไฟฟ้านครหลวงเขตธนบุรี
2. การไฟฟ้านครหลวงเขตนนทบุรี
3. การไฟฟ้านครหลวงเขตนวลจันทร์
4. การไฟฟ้านครหลวงเขตบางกะปิ
5. การไฟฟ้านครหลวงเขตบางขุนเทียน
6. การไฟฟ้านครหลวงเขตบางนา
7. การไฟฟ้านครหลวงเขตบางบัวทอง
8. การไฟฟ้านครหลวงเขตบางพลี
9. การไฟฟ้านครหลวงเขตบางเขน
10. การไฟฟ้านครหลวงเขตบางใหญ่
11. การไฟฟ้านครหลวงเขตมีนบุรี
12. การไฟฟ้านครหลวงเขตยานนาวา
13. การไฟฟ้านครหลวงเขตราษฎร์บูรณะ
14. การไฟฟ้านครหลวงเขตลาดกระบัง
15. การไฟฟ้านครหลวงเขตวัดเลียบ
16. การไฟฟ้านครหลวงเขตสมุทรปราการ
17. การไฟฟ้านครหลวงเขตสามเสน
18. การไฟฟ้านครหลวงสำนักงานใหญ่ (เขตคลองเตย)

ครอบคลุมพื้นที่ 3 จังหวัด ได้แก่ กรุงเทพมหานคร นนทบุรี และสมุทรปราการ

---



## **1. Setup & Load Data**

In [1]:
# import relevant libralies
import pandas as pd
import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
from statsmodels.tsa.seasonal import seasonal_decompose
import warnings
warnings.filterwarnings('ignore')

In [2]:
# read data from file path using 'utf-8-sig' for thai alphabets
filelink = "https://storagepracticebdidev.blob.core.windows.net/dev-uploads/app/bdi/uploads1-9-20260312-162658503.csv?sv=2026-02-06&st=2026-03-12T09%3A26%3A58Z&se=2028-12-06T09%3A26%3A58Z&sr=b&sp=r&sig=fV6wspFvE9dTgkzY9xLbZ%2FXP4i%2B%2FN0BliGZAx9%2Bzywo%3D"
df = pd.read_csv(filelink, encoding='utf-8-sig')

In [3]:
#ดาวน์โหลดฟอนต์ภาษาไทย (TH Sarabun New)
!wget -q https://github.com/Phonbopit/sarabun-webfont/raw/master/fonts/thsarabunnew-webfont.ttf
#เพิ่มฟอนต์เข้าไปในระบบ
mpl.font_manager.fontManager.addfont('thsarabunnew-webfont.ttf')
mpl.rc('font', family='TH Sarabun New', size=16)
sns.set_theme(style="whitegrid", font="TH Sarabun New")

# ตั้งค่า style
plt.rcParams['figure.dpi'] = 120

In [4]:
# # connect colab environment to google drive
# from google.colab import drive
# drive.mount('/content/drive')

# # read data from file path using 'utf-8-sig' for thai alphabets
# filepath = "/content/drive/MyDrive/SuperAIengineer6/mini-Hackathon_1/E68.csv"
# df = pd.read_csv(filepath, encoding='utf-8-sig')

## **2. First Look and Data Cleaning**

### 2.1 ตรวจสอบภาพรวมข้อมูล

In [5]:
# Check overview information of the data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 83 entries, 0 to 82
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   ปี             7 non-null      float64
 1   เดือน          83 non-null     object 
 2    วัดเลียบ      83 non-null     float64
 3    คลองเตย       83 non-null     float64
 4    ยานนาวา       83 non-null     float64
 5    บางกะปิ       83 non-null     float64
 6    มีนบุรี       83 non-null     float64
 7    สมุทรปราการ   83 non-null     float64
 8    บางพลี        83 non-null     float64
 9    สามเสน        83 non-null     float64
 10   นนทบุรี       83 non-null     float64
 11   บางใหญ่       83 non-null     float64
 12   ธนบุรี        83 non-null     float64
 13   ราษฎร์บูรณะ   83 non-null     float64
 14   บางขุนเทียน   83 non-null     float64
 15   บางเขน        83 non-null     float64
 16   บางนา         83 non-null     float64
 17   บางบัวทอง     83 non-null     float64
 18   ลาดกระบัง  

In [6]:
# check the data shape
df.shape

(83, 21)

In [7]:
# check some examples of the data
df.head(13)

,ปี,เดือน,วัดเลียบ,คลองเตย,ยานนาวา,บางกะปิ,มีนบุรี,สมุทรปราการ,บางพลี,สามเสน,...,บางใหญ่,ธนบุรี,ราษฎร์บูรณะ,บางขุนเทียน,บางเขน,บางนา,บางบัวทอง,ลาดกระบัง,นวลจันทร์,(หักไฟทำการ)
0,2562.0,ม.ค.,93.96,375.86,154.62,243.49,156.70,400.83,409.56,291.46,...,134.90,152.69,321.13,199.88,239.95,220.28,146.70,152.37,150.26,"4,119.31"
1,NaN,ก.พ.,94.88,373.54,156.55,248.90,161.21,396.36,408.68,298.58,...,142.73,158.55,314.98,202.95,254.82,222.29,152.47,155.39,158.49,"4,188.50"
2,NaN,มี.ค.,103.27,412.09,170.67,272.89,180.84,445.57,458.59,328.44,...,159.21,174.84,361.92,225.96,281.99,246.60,172.70,167.90,174.00,"4,660.94"
3,NaN,เม.ย.,108.59,419.71,179.36,293.99,188.01,400.70,426.82,345.59,...,173.20,187.59,338.99,224.96,295.08,255.97,176.43,172.99,187.62,"4,709.44"
4,NaN,พ.ค.,110.97,429.76,184.89,307.18,195.85,446.77,468.74,355.00,...,176.45,193.47,368.12,243.04,307.90,268.79,186.15,194.68,193.52,"4,976.37"
5,NaN,มิ.ย.,102.74,406.44,171.90,282.80,177.51,430.97,446.09,329.44,...,159.98,178.04,354.57,227.89,276.05,252.02,171.66,172.28,174.60,"4,638.70"
6,NaN,ก.ค.,100.45,400.66,166.92,270.44,175.60,432.00,442.83,317.67,...,155.39,173.00,341.48,219.75,270.36,243.73,168.35,165.00,168.99,"4,524.05"
7,NaN,ส.ค.,101.31,402.03,166.13,271.31,173.87,423.35,442.82,325.68,...,153.30,172.12,347.63,218.46,270.70,243.95,169.11,165.27,167.26,"4,524.33"
8,NaN,ก.ย.,94.45,382.51,156.76,259.60,168.06,408.03,434.78,300.67,...,144.60,162.87,328.64,206.91,257.68,234.13,158.86,157.57,158.74,"4,309.26"
9,NaN,ต.ค.,98.84,402.54,166.13,266.20,176.88,417.81,451.06,324.00,...,149.95,168.61,334.34,211.80,269.58,245.71,167.18,166.56,166.53,"4,489.65"


จากการตรวจสอบข้อมูลพบว่ามีทั้งหมดข้อมูลทั้ง 83 แถวแบ่งเป็น เดือนในปี 2562-2568 (7 ปี) ที่เก็บข้อมูลได้ โดยในปี 2568 เก็บข้อมูลถึงเดือน พ.ย (12x6+11 =83) และข้อมูลทั้ง 21 คอลัมน์ ประกอบไปด้วย
1. ปี
2. เดือน
3. รายชื่อทั้ง 18 เขตทำการของการไฟฟ้านครหลวง
4. (หักไฟทำการ)

โดยจากการตรวจสอบข้อมูลคร่าวๆพบว่า
1. ข้อมูลปีหายไป
2. ข้อมูลเดือนต้องตรวจสอบจำนวนรูปแบบเพิ่มเติม
3. ข้อมูลคอลัมน์ (หักไฟทำการ) ที่ไม่พบคำอธิบายรายละเอียด และประเภทข้อมูลเป็น object type

### 2.2 ตรวจสอบ missing values
เช่น ข้อมูลปีของแต่ละแถว และทำการใส่ข้อมูลที่ถูกต้องลงไป

In [8]:
# check if only year is missing
df.isnull().sum()

ปี               76
เดือน             0
 วัดเลียบ         0
 คลองเตย          0
 ยานนาวา          0
 บางกะปิ          0
 มีนบุรี          0
 สมุทรปราการ      0
 บางพลี           0
 สามเสน           0
 นนทบุรี          0
 บางใหญ่          0
 ธนบุรี           0
 ราษฎร์บูรณะ      0
 บางขุนเทียน      0
 บางเขน           0
 บางนา            0
 บางบัวทอง        0
 ลาดกระบัง        0
 นวลจันทร์        0
(หักไฟทำการ)      0
dtype: int64

In [9]:
# check how year is missing
df.dropna()

,ปี,เดือน,วัดเลียบ,คลองเตย,ยานนาวา,บางกะปิ,มีนบุรี,สมุทรปราการ,บางพลี,สามเสน,...,บางใหญ่,ธนบุรี,ราษฎร์บูรณะ,บางขุนเทียน,บางเขน,บางนา,บางบัวทอง,ลาดกระบัง,นวลจันทร์,(หักไฟทำการ)
0,2562.0,ม.ค.,93.96,375.86,154.62,243.49,156.70,400.83,409.56,291.46,...,134.90,152.69,321.13,199.88,239.95,220.28,146.70,152.37,150.26,"4,119.31"
12,2563.0,ม.ค.,98.48,395.39,161.56,264.90,166.61,391.37,419.47,314.79,...,147.00,166.19,327.96,206.00,260.64,233.28,160.33,157.25,160.86,"4,330.05"
24,2564.0,ม.ค.,70.45,277.03,125.05,201.14,142.35,358.71,379.94,233.44,...,123.50,135.15,285.55,179.41,210.03,193.55,146.15,135.90,125.64,"3,568.79"
36,2565.0,ม.ค.,76.13,310.65,141.02,230.48,162.77,395.97,415.39,260.24,...,143.74,154.78,312.42,200.71,238.13,220.51,170.38,156.45,144.29,"4,007.87"
48,2566.0,ม.ค.,76.43,327.29,134.72,223.13,149.93,358.33,396.69,254.48,...,130.11,142.38,273.83,182.47,230.69,207.04,158.52,152.40,130.28,"3,783.25"
60,2567.0,ม.ค.,90.13,380.42,157.02,266.80,175.39,376.18,434.64,302.72,...,159.06,169.25,301.50,206.76,272.62,241.85,187.77,168.37,155.26,"4,341.56"
72,2568.0,ม.ค.,77.57,354.62,138.05,236.56,149.30,346.52,394.81,265.04,...,133.06,143.92,268.20,178.52,233.34,209.68,166.48,143.23,129.95,"3,825.42"


จะเห็นได้ว่ามีแค่ข้อมูลปีที่หายไป โดยข้อมูลที่ได้ มีการใส่เลขปีแค่ที่เดือนม.ค. จึงทำการเพิ่มปีข้อมูลปีใน dataset โดยอ้างอิงตามเลขปีที่ใส่ในเดือน ม.ค. ก่อนหน้า

In [10]:
# use forward fill (ffill) to handle missing data by propagating the last valid forward to the next valid values

df['ปี'] = df['ปี'].ffill()
df

,ปี,เดือน,วัดเลียบ,คลองเตย,ยานนาวา,บางกะปิ,มีนบุรี,สมุทรปราการ,บางพลี,สามเสน,...,บางใหญ่,ธนบุรี,ราษฎร์บูรณะ,บางขุนเทียน,บางเขน,บางนา,บางบัวทอง,ลาดกระบัง,นวลจันทร์,(หักไฟทำการ)
0,2562.0,ม.ค.,93.96,375.86,154.62,243.49,156.70,400.83,409.56,291.46,...,134.90,152.69,321.13,199.88,239.95,220.28,146.70,152.37,150.26,"4,119.31"
1,2562.0,ก.พ.,94.88,373.54,156.55,248.90,161.21,396.36,408.68,298.58,...,142.73,158.55,314.98,202.95,254.82,222.29,152.47,155.39,158.49,"4,188.50"
2,2562.0,มี.ค.,103.27,412.09,170.67,272.89,180.84,445.57,458.59,328.44,...,159.21,174.84,361.92,225.96,281.99,246.60,172.70,167.90,174.00,"4,660.94"
3,2562.0,เม.ย.,108.59,419.71,179.36,293.99,188.01,400.70,426.82,345.59,...,173.20,187.59,338.99,224.96,295.08,255.97,176.43,172.99,187.62,"4,709.44"
4,2562.0,พ.ค.,110.97,429.76,184.89,307.18,195.85,446.77,468.74,355.00,...,176.45,193.47,368.12,243.04,307.90,268.79,186.15,194.68,193.52,"4,976.37"
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
78,2568.0,ก.ค.,97.02,419.06,171.30,295.84,197.27,421.38,482.65,328.03,...,181.01,188.45,322.13,222.60,306.31,269.26,216.13,177.19,173.98,"4,800.15"
79,2568.0,ส.ค.,98.48,429.89,170.95,305.49,199.75,414.64,485.13,336.91,...,185.54,193.74,319.74,225.90,313.56,274.17,219.57,188.20,179.39,"4,881.46"
80,2568.0,ก.ย.,92.91,408.87,163.69,287.20,188.05,410.30,475.33,318.31,...,172.17,181.35,309.14,213.33,289.19,262.69,207.16,174.44,164.13,"4,635.66"
81,2568.0,ต.ค.,91.96,413.57,162.70,284.44,187.98,411.02,472.91,314.98,...,171.22,178.09,315.09,211.28,292.78,259.38,211.05,174.31,163.81,"4,635.88"


In [11]:
# recheck if there is still missing value
df.isnull().sum()

ปี               0
เดือน            0
 วัดเลียบ        0
 คลองเตย         0
 ยานนาวา         0
 บางกะปิ         0
 มีนบุรี         0
 สมุทรปราการ     0
 บางพลี          0
 สามเสน          0
 นนทบุรี         0
 บางใหญ่         0
 ธนบุรี          0
 ราษฎร์บูรณะ     0
 บางขุนเทียน     0
 บางเขน          0
 บางนา           0
 บางบัวทอง       0
 ลาดกระบัง       0
 นวลจันทร์       0
(หักไฟทำการ)     0
dtype: int64

ตรวจสอบพบว่าไม่มี missing values แล้ว จากนั้นเราจะทำการสร้างคอลัมน์ปี ค.ศ. ขึ้นมาเพื่อง่ายต่อการนำไปใช้ plot time series

In [12]:
df['Year'] = (df['ปี']-543).astype(int)
df['Year'].unique()

array([2019, 2020, 2021, 2022, 2023, 2024, 2025])

### 2.3 ตรวจสอบข้อมูลคอลัมน์ "เดือน"

In [13]:
# check unique of the column
df['เดือน'].unique()

array([' ม.ค. ', ' ก.พ. ', ' มี.ค. ', ' เม.ย. ', ' พ.ค. ', ' มิ.ย. ',
       ' ก.ค. ', ' ส.ค. ', ' ก.ย. ', ' ต.ค. ', ' พ.ย. ', ' ธ.ค. ',
       '\xa0ม.ค.\xa0', '\xa0ก.พ.\xa0', ' \xa0มี.ค.\xa0 ',
       ' \xa0เม.ย.\xa0 ', 'ต.ค.', 'พ.ย.', 'ธ.ค.', 'ม.ค.', 'ก.พ.', 'มี.ค.',
       'เม.ย.', 'พ.ค.', 'มิ.ย.', 'ก.ค.', 'ส.ค.', 'ก.ย.'], dtype=object)

จะเห็นว่าข้อมูลที่ควรจะหมายถึงเดือนเดียวกัน เช่น ' ม.ค. ' '\xa0ม.ค.\xa0' และ 'ม.ค.' แบ่งออกเป็น 3 แบบ ขั้นตอนต่อไปนี้จะต้องทำให้เหลือแบบเดียว เพื่อง่ายต่อการประมวลผลต่อไป

In [14]:
# replace '\xa0' with empty
df['เดือน'] = df['เดือน'].str.replace('\xa0', '')
df['เดือน'].unique()

array([' ม.ค. ', ' ก.พ. ', ' มี.ค. ', ' เม.ย. ', ' พ.ค. ', ' มิ.ย. ',
       ' ก.ค. ', ' ส.ค. ', ' ก.ย. ', ' ต.ค. ', ' พ.ย. ', ' ธ.ค. ', 'ม.ค.',
       'ก.พ.', 'ต.ค.', 'พ.ย.', 'ธ.ค.', 'มี.ค.', 'เม.ย.', 'พ.ค.', 'มิ.ย.',
       'ก.ค.', 'ส.ค.', 'ก.ย.'], dtype=object)

In [15]:
# replace a space ' ' with empty
df['เดือน'] = df['เดือน'].str.strip()
df['เดือน'].unique()

array(['ม.ค.', 'ก.พ.', 'มี.ค.', 'เม.ย.', 'พ.ค.', 'มิ.ย.', 'ก.ค.', 'ส.ค.',
       'ก.ย.', 'ต.ค.', 'พ.ย.', 'ธ.ค.'], dtype=object)

จะเห็นว่าตอนนี้ข้อมูลในคอลัมน์เดือน ชื่อแต่ละเดือนมีรูปแบบเดียว สามารถนำไปใช้วิเคราะห์ผลได้ เช่น การหาผลรวมของจำนวนหน่วยไฟฟ้าจำแนกตามเดือน

ขั้นตอนต่อไปจะเป็นการ map ชื่อเดือนเป็นตัวเลข เช่น ม.ค. เป็น 1 โดยสร้างอีกคอลัมน์ขึ้นมา

In [16]:
thai_month_map = {
    'ม.ค.': 1, 'ก.พ.': 2, 'มี.ค.': 3,
    'เม.ย.': 4, 'พ.ค.': 5, 'มิ.ย.': 6,
    'ก.ค.': 7, 'ส.ค.': 8, 'ก.ย.': 9,
    'ต.ค.': 10, 'พ.ย.': 11, 'ธ.ค.': 12
}

df['Month'] = df['เดือน'].map(thai_month_map)
df['Month'].unique()

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12])

### 2.4 ตรวจสอบข้อมูลคอลัมน์ "(หักไฟทำการ)"

เนื่องจากประเภทข้อมูลเป็น object และไม่พบคำอธิบายรายละเอียด ซึ่งสันนิษฐานว่าคือข้อมูลจำนวนหน่วยไฟฟ้ารวมที่การไฟฟ้านครหลวงจำหน่ายไปในแต่ละเดือน

โดยในขั้นตอนนี้จะทำการแปลงข้อมูลและตรวจสอบว่าใช่ตามข้อสันนิษฐานหรือไม่

In [17]:
# change type of the data from object to float64
df['(หักไฟทำการ)']= df['(หักไฟทำการ)'].str.replace(',','').astype(float)

In [18]:
# recheck if the data type was changed to float
df.dtypes

ปี               float64
เดือน             object
 วัดเลียบ        float64
 คลองเตย         float64
 ยานนาวา         float64
 บางกะปิ         float64
 มีนบุรี         float64
 สมุทรปราการ     float64
 บางพลี          float64
 สามเสน          float64
 นนทบุรี         float64
 บางใหญ่         float64
 ธนบุรี          float64
 ราษฎร์บูรณะ     float64
 บางขุนเทียน     float64
 บางเขน          float64
 บางนา           float64
 บางบัวทอง       float64
 ลาดกระบัง       float64
 นวลจันทร์       float64
(หักไฟทำการ)     float64
Year               int64
Month              int64
dtype: object

จะเห็นได้ว่า datatype ถูกเปลี่ยนเป็น float64 แล้ว สามารถใช้เปรียบเทียบข้อมูลได้

In [19]:
# create office_name lists from column names
column_names = df.columns
office_names = list(column_names.drop(['ปี', 'เดือน', '(หักไฟทำการ)','Year', 'Month']))
office_names

[' วัดเลียบ ',
 ' คลองเตย ',
 ' ยานนาวา ',
 ' บางกะปิ ',
 ' มีนบุรี ',
 ' สมุทรปราการ ',
 ' บางพลี ',
 ' สามเสน ',
 ' นนทบุรี ',
 ' บางใหญ่ ',
 ' ธนบุรี ',
 ' ราษฎร์บูรณะ ',
 ' บางขุนเทียน ',
 ' บางเขน ',
 ' บางนา ',
 ' บางบัวทอง ',
 ' ลาดกระบัง ',
 ' นวลจันทร์ ']

In [20]:
# calculate sum of values in 'office_name' column and
total_unit = df[office_names].sum(axis=1)

# ccompare to the last column if the difference less than 0.1
compare_lists = list(df['(หักไฟทำการ)']-total_unit < 0.1)
compare_lists.count(False)

0

จากการตรวจสอบพบว่าข้อมูลในคอลัมน์สุดท้าย (หักไฟทำการ) มีเท่ากับจำนวนหน่วยไฟฟ้ารวมทั้งหมดที่จำหน่ายในแต่ละเดือนของการไฟฟ้านครหลวง (ต่างกันไม่เกิน 0.1) จึงสรุปว่าข้อมูลในคอลัมน์นี้คือ หน่วยจำหน่ายไฟฟ้ารวมในแต่ละเดือน

In [21]:
# change name of the last column
df = df.rename(columns={'(หักไฟทำการ)': 'Total_Consumption'})
df.head()

,ปี,เดือน,วัดเลียบ,คลองเตย,ยานนาวา,บางกะปิ,มีนบุรี,สมุทรปราการ,บางพลี,สามเสน,...,ราษฎร์บูรณะ,บางขุนเทียน,บางเขน,บางนา,บางบัวทอง,ลาดกระบัง,นวลจันทร์,Total_Consumption,Year,Month
0,2562.0,ม.ค.,93.96,375.86,154.62,243.49,156.70,400.83,409.56,291.46,...,321.13,199.88,239.95,220.28,146.70,152.37,150.26,4119.31,2019,1
1,2562.0,ก.พ.,94.88,373.54,156.55,248.90,161.21,396.36,408.68,298.58,...,314.98,202.95,254.82,222.29,152.47,155.39,158.49,4188.50,2019,2
2,2562.0,มี.ค.,103.27,412.09,170.67,272.89,180.84,445.57,458.59,328.44,...,361.92,225.96,281.99,246.60,172.70,167.90,174.00,4660.94,2019,3
3,2562.0,เม.ย.,108.59,419.71,179.36,293.99,188.01,400.70,426.82,345.59,...,338.99,224.96,295.08,255.97,176.43,172.99,187.62,4709.44,2019,4
4,2562.0,พ.ค.,110.97,429.76,184.89,307.18,195.85,446.77,468.74,355.00,...,368.12,243.04,307.90,268.79,186.15,194.68,193.52,4976.37,2019,5


In [22]:
# last check the dataset before the next processes
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 83 entries, 0 to 82
Data columns (total 23 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   ปี                 83 non-null     float64
 1   เดือน              83 non-null     object 
 2    วัดเลียบ          83 non-null     float64
 3    คลองเตย           83 non-null     float64
 4    ยานนาวา           83 non-null     float64
 5    บางกะปิ           83 non-null     float64
 6    มีนบุรี           83 non-null     float64
 7    สมุทรปราการ       83 non-null     float64
 8    บางพลี            83 non-null     float64
 9    สามเสน            83 non-null     float64
 10   นนทบุรี           83 non-null     float64
 11   บางใหญ่           83 non-null     float64
 12   ธนบุรี            83 non-null     float64
 13   ราษฎร์บูรณะ       83 non-null     float64
 14   บางขุนเทียน       83 non-null     float64
 15   บางเขน            83 non-null     float64
 16   บางนา             83 non-nu

ขั้นตอนของ data cleaning สำเร็จแล้ว ✅

## **3. Basic EDA**

### 3.1 Statistical summary


In [23]:
# Generates descriptive statistics of the dataset
df.describe().T

,count,mean,std,min,25%,50%,75%,max
ปี,83.0,2564.963855,1.996618,2562.00,2563.000,2565.00,2567.000,2568.00
วัดเลียบ,83.0,89.483253,9.149820,68.39,83.675,90.13,95.940,110.97
คลองเตย,83.0,366.394699,44.620443,276.10,331.460,373.54,405.455,441.01
ยานนาวา,83.0,157.913012,13.521175,125.05,147.170,157.52,165.995,187.52
บางกะปิ,83.0,263.637590,27.462968,201.14,247.545,262.05,286.160,328.15
มีนบุรี,83.0,178.507349,16.871710,142.35,167.360,177.51,191.320,218.25
สมุทรปราการ,83.0,401.308072,24.879527,346.52,385.270,405.40,417.500,446.77
บางพลี,83.0,439.188916,31.226581,376.73,414.305,441.23,463.845,509.68
สามเสน,83.0,301.794458,28.840888,233.44,280.655,302.72,323.730,361.50
นนทบุรี,83.0,305.446265,26.760353,245.82,287.425,306.98,323.590,368.02


จากข้อมูลจะเห็นว่าค่าเฉลี่ยจำนวนหน่วยจำหน่ายไฟฟ้าตลอดระยะเวลาการเก็บข้อมูลปี 2562-2568 สูงสุด 3 อันดับ คือ
1. เขตบางพลี ประมาณ 439 ล้านหน่วยต่อเดือน
2. เขตสมุทรปราการ ประมาณ 401 ล้านหน่วยต่อเดือน
3. เขตคลองเตย ประมาณ 366 ล้านหน่วยต่อเดือน

โดยเขตคลองเตยมีค่าส่วนเบี่ยงเบนมาตรฐานสูงสุด ซึ่งหมายถึงมีความแตกต่างของจำนวนหน่วยไฟฟ้าที่จำหน่ายในแต่ละเดือนสูงสุด โดยจะต้องพิจารณาต่อไป ทั้งนี้ อาจสามารถอนุมานได้ว่าเขตที่มีจำนวนโรงงานอุตสาหกรรมมาก คือ เขตบางพลีและสมุทรปราการ ซึ่งมีการใช้ไฟฟ้ามากกว่า 400 ล้านหน่วยต่อเดือน

และเขตที่มีค่าเฉลี่ยจำนวนหน่วยไฟฟ้าต่ำสุด 3 อันดับคือ วัดเลียบ ยานนาวา บางใหญ่ ซึ่งอาจเป็นเขตชุมชนอยู่อาศัย


---


ขั้นตอนต่อไปจะทดลองพล็อตกราฟเพื่อดูแนวโน้มของข้อมูล โดยเริ่มจากการสร้างคอลัมน์ Date ขึ้นมา และใช้วันเป็นวันที่ 1 สำหรับทุกแถวข้อมูล

จากนั้นจะใช้ Date เป็น index ของข้อมูลเพื่อความสะดวก

In [24]:
# Create a column 'date' from 'month' and 'year' for plotting
df['Date'] = pd.to_datetime(df[['Month', 'Year']].assign(Day=1))
df.set_index('Date', inplace = True)

### 3.2 Multiple-line plot
ขั้นตอนนี้จะเป้นการดูภาพรวมการใช้พลังงานแยกแต่ละเขต (มาก/น้อย) ของแต่ละเดือนในปี ตลอดระยะเวลาเก็บข้อมูล ปี 2562-2568

In [25]:
# 1.Use plotly express to plot all the data
fig = px.line(df,
              x= df.index,
              y= office_names,
              title='Electricity Consumption Trend (2018-2025)',
              labels={'value': 'Electricity Consumption (GWh)', 'date_col': 'Date'},
              template='plotly_white',
              color_discrete_sequence=px.colors.qualitative.Alphabet)

# Improve the layout
# 2. Put legend at the bootom of the plot
fig.update_layout(
    legend=dict(
        orientation="h",  # Horizontal orientation
        yanchor="bottom",
        y=-0.3,           # Places it below the X-axis
        xanchor="center",
        x=0.5
    )
)
fig.show()

จาก Multiple-line plot จะเห็นภาพรวมการใช้พลังงานของทั้ง 18 เขตคล้ายกับ statistical overview ในข้อ 3.1 โดยเราจะสามารถแบ่งกลุ่มการใช้พลังงานมากน้อยเป็น 4 กลุ่มใหญ่ๆ คือ
1. เขตที่ใช้พลังงานมาก 3 เขต คือประมาณ 400 GWh/เดือน คือ บางพลี สมุทรปราการ คลองเตย
2. เขตที่ใช้พลังงานประมาณ 200-300 GWh/เดือน 7 เขต คึอ ราษฎร์บูรณะ สามเสน นนทบุรี บางกะปี บางเขน บางนา บางขุนเทียน
3. เขตที่ใช้พลังงานประมาณ 100-200 GWh/เดือน 7 เขต คึอ มีนบุรี บางบัวทอง ธนบุรี บางใหญ่ ลาดกระบัง นวลจันทร์ ยานนาวา
4. เขตที่ใช้พลังงานน้อย 1 เขต คือ ประมาณ 100 GWh/เดือน หรือน้อยกว่า คือ วัดเลียบ

โดยจากกราฟยังมีข้อสังเกตเพิ่มเติมของการใช้ไฟฟ้าบางเขตและความผิดปกติของข้อมูลบางจุด ดังนี้
1. เขตคลองเตย มีการใช้ไฟฟ้าสูงในปี 2562(2019) ถึงต้นปี 2563(2020) และมีช่วงระยะเวลาที่ใช้ไฟฟ้าลดลงในปี 2563 ถึงต้นปี 2565(2022) ก่อนจะกลับมาใช้ไฟฟ้าในปริมาณเดิมในช่วงก่อนหน้า
2. เขตบางบัวทอง มีปริมาณการใช้ไฟฟ้าต่อเดือนอยู่ในช่วง 200 GWh ในเดือน พ.ค. หรือน้อยกว่าตลอดทั้งปี จนถึงช่วงต้นปี 2566 (2023) เป็นต้นมา ที่มีบางเดือนใช้ไฟฟ้าสูงขึ้นเป็น 240 GWh เช่นในเดือนพ.ค. 2567 คิดเป็น +20% โดยมีแนวโน้มโดยรวมสูงขึ้นตลอดทั้งปีด้วย  
3. มีความเป็นไปได้ที่ข้อมูลหน่วยจำหน่ายไฟฟ้าในเดือนกันยายนปี 2563(2020) จะสลับกันระหว่างเขตลาดกระบังและเขตบางนา



---



## **4. โจทย์วิเคราะห์ข้อมูล (Challenge Questions)**🧠


###**ข้อ 1 — การใช้ไฟฟ้ารวมเปลี่ยนแปลงอย่างไรใน 7 ปี?**

Trend & Anomaly Detection

*โจทย์*
จากข้อมูลหน่วยจำหน่ายไฟฟ้ารวม **รายเดือนตั้งแต่ปี 2562–2568**
จงสร้างกราฟที่แสดง
* แนวโน้ม (Trend) การใช้ไฟฟ้าโดยรวม
* รูปแบบฤดูกาล (Seasonality)
* และตรวจจับ จุดผิดปกติ (Anomaly) ที่เบี่ยงออกจากรูปแบบปกติ

พร้อมระบุให้ได้ว่าเหตุการณ์ใดน่าจะอธิบายจุดเหล่านั้น เช่น
* การระบาดของ COVID-19
* คลื่นความร้อน
* มาตรการหรือนโยบายของรัฐ

**เป้าหมายของโจทย์**
ต้องสามารถสื่อสารได้ว่า แนวโน้มการใช้ไฟฟ้าในปี 2568 อยู่ในทิศทางใด และสะท้อน ภาวะเศรษฐกิจเมืองอย่างไร

---

ในข้อนี้จะใช้ seasonal_decompose ของ time-series analysis จาก statsmodels มีขั้นตอนดังนี้
1. **DECOMPOSITION**: extract 3 components จากข้อมูลการใช้ไฟฟ้ารวมในคอลัมน์ Total_Consumption โดยใช้ additive model โดย assume ว่าข้อมูลของเรา  Raw_data[t] = Trend[t] + Seasonal[t] + Residual[t]
2. **ANOMALY DETECTION**: หา anomalies ของข้อมูลโดยใช้ Residual ที่ได้ เพื่อดูว่า จุดไหนค่า Residual มากกว่า 2 เท่าของ std
3. **VISUALIZATION**: plot ค่า trend และ seasonality ที่ decompose ได้ เทียบกับข้อมูลดิบ
4. **ANNOTATING HISTORICAL EVENTS**: ระบุเหตุการณ์สำคัญจากตัวอย่างเพิ่มลงไปในกราฟ เช่น covid-19 และ heat wave

In [26]:
# 1. DECOMPOSITION (Trend & Seasonality)
# Period=12 because the data is monthly
decomposition = seasonal_decompose(df['Total_Consumption'],
                                   model='additive', period=12)

df['Trend'] = decomposition.trend
df['Seasonal'] = decomposition.seasonal
df['Residual'] = decomposition.resid

# 2. ANOMALY DETECTION
# Using Residuals: anomalies are identified if the 'Residual' (noise) > 2*std
std_dev = df['Residual'].std()
df['Is_Anomaly'] = (df['Residual'].abs() > (2*std_dev))
anomalies = df[df['Is_Anomaly'] == True]


# 3. VISUALIZATION
fig = go.Figure()

# Plot the Actual Data (Raw)
fig.add_trace(go.Scatter(x=df.index, y=df['Total_Consumption'],
                         name='Actual Consumption', line=dict(color='lightgrey',
                                                              width=1)))

# Plot the Trend (smooth direction)
fig.add_trace(go.Scatter(x=df.index, y=df['Trend'],
                         name='Trend', line=dict(color='cyan', width=4)))

# Plot the "Expected" peaks every year (seasonality)
fig.add_trace(go.Scatter(x=df.index, y=df['Total_Consumption']-df['Seasonal'],
                         name='Seasonal Pattern', line=dict(color='orange',
                                                            width=1, dash='dot')))

# Highlight Anomalies (Red Dots)
fig.add_trace(go.Scatter(x=anomalies.index, y=anomalies['Total_Consumption'],
                         mode='markers', name='Anomalies',
                         marker=dict(color='red', size=10, symbol='x')))

# 4. ANNOTATING HISTORICAL EVENTS
events = [
    dict(date="2020-01-01", label="COVID-19 First Case"),
    dict(date="2020-04-01", label="COVID-19 Lock Down"),
    dict(date="2024-05-01", label="Extreme Heat Wave"),
]

for event in events:
    if pd.to_datetime(event['date']) in df.index:
        fig.add_annotation(x=event['date'], y=df.loc[event['date'], 'Total_Consumption'],
                           text=event['label'], showarrow=True, arrowhead=2)

fig.update_layout(
    title='<b>Electricity Consumption:</b> Trend, Seasonality, and Anomalies (2018-2025)',
    xaxis_title='Year',
    yaxis_title='Electricity Consumption (GWh)',
    template='plotly_dark',
    hovermode='x unified'
)

fig.show()

* จากกราฟจะเห็นได้ว่าแนวโน้มการใช้ไฟฟ้าโดยรวมตลอดช่วงปี 2563-2564(2020-2021) ลดลงจากปี 2562 (2019) และค่อนข้างคงที่ จากนั้นเพิ่มสูงขึ้นในปี 2565-2567(2022-2024) ช่วงกลางปี ก่อนจะ**ลดลงอีกครั้งในช่วงปลายปี 2567(2024) ถึงปี 2568 (2025)** ซึ่งการลดลงของการใช้ไฟฟ้าอาจแสดงถึง**ภาวะเศรษฐกิจหดตัวลง (Economic Contraction)** ในปี 2568 ทำให้แนวโน้มการใช้ไฟฟ้าโดยรวม ซึ่งอาจรวมถึงการใช้ไฟฟ้าสำหรับกิจกรรมทางเศรษฐกิจและอุตสาหกรรมนั้นลดลงไปด้วย

* จากการตรวจสอบ**จุดผิดปกติ (anmalies)** พบ 3 จุด มี 2 จุดอยู่ในปี 2563 (2020) ซึ่งระบุได้ว่าเป็นผลมาจากการระบาดและมาตรการเกี่ยวกับ COVID-19 โดยนอกจากนี้การระบาดของ COVID-19 ยังทำให้แนวโน้มการใช้ไฟฟ้าหลังจากนั้นจนถึงปี 2565 (2022) ลดลงและคงที่ เนื่องจากผลของเศรษฐกิจหดตัวด้วย โดยหากย้อนกลับไปที่ multiple-line plot ในข้อ 3.2 จะเห็นผลของ COVID-19 นี้ชัดเจนที่สุดที่เขตคลองเตย (ตามที่ได้อธิบายไปก่อนหน้านี้) ซึ่งสันนิษฐานสาเหตุที่เขตคลองเตยได้รับผลกระทบจากเรื่องนี้มากที่สุดเนื่องจาก ในเขตคลองเตยมีท่าเรือคลองเตย หรือท่าเรือกรุงเทพ ซึ่งเป็นท่าเรือสำคัญของกรุงเทพมหานครสำหรับขนส่งสินค้าคอนเทนเนอร์และสินค้าทั่วไป อาจจะได้รับผลกระทบจากเศรษฐกิจ ทำให้ใช้ไฟฟ้าน้อยลงอย่างเห็นได้ชัดในช่วงเวลาดังกล่าว

* อีกจุดผิดปกติอยู่ที่เดือนมกราคมปี 2568 (2025) ที่มีการใช้ไฟฟ้าต่ำกว่าปกติ โดยสาเหตุยังไม่สามารถระบุได้แน่ชัด โดยหากย้อนกลับไปดูที่ multiple-line plot ในข้อ 3.2 จะเห้นว่าทุกเขตมีการใช้ไฟฟ้าลดลงในช่วงเวลานี้ด้วย

* ส่วนเหตุการณ์ heat wave ในเดือนพฤษภาคม ปี 2567 (2024) จะเห็นว่าในเดือนดังกล่าวมีปริมาณการใช้ไฟฟ้ารวมสูงสุด อย่างไรก็ตามเมื่อพิจารณาค่าที่เบี่ยงเบนออกจากแนวโน้มโดยรวมนั้น จะไม่ถูกระบุว่าเป็นค่าที่ผิดปกติ หรือ anomalies

---



### **ข้อ 2 — เขตไหนใช้ไฟมากที่สุด และใครโตเร็วที่สุด?**

District Ranking & Growth

*โจทย์* จากข้อมูลของ 18 เขตการจำหน่ายไฟฟ้า จงสร้างกราฟที่แสดง
* ปริมาณการใช้ไฟฟ้า และ
* อัตราการเติบโตของการใช้ไฟฟ้าของแต่ละเขต **พร้อมกันในกราฟเดียว**

เพื่อระบุให้ได้ว่า
* เขตใดคือ เขตที่ใช้ไฟมากและยังโตต่อ
* เขตใดคือ ใช้น้อยแต่โตเร็ว
ซึ่งอาจสะท้อน **พื้นที่เมืองที่กำลังขยายตัว**
---

ในข้อนี้จะทำการสร้างกราฟที่แสดงปริมาณการใช้ไฟฟ้ารวมในปี 2568 (2025) และอัตราการเติบโตของการใช้ไฟฟ้าของแต่ละเขตเมื่อเทียบกับปี 2567(2024) เพื่อทำ Quadrant Analysis โดยมีขั้นตอนดังนี้
1. **PREPARE DATA**: เลือกข้อมูลเฉพาะช่วงปี 2025 และ 2024 ของแต่ละเขตเก็บเป็น dataframe แยกกัน โดยใช้ข้อมูลของเดือนมกราคม-พฤศจิกายน เนื่องจากไม่มีข้อมูลเดือนธันวาคม ของปี 2025 (เพื่อการเปรียบเทียบที่เท่ากัน) จากนั้นสร้าง dataframe เก็บข้อมูลของปี 2025 และคำนวณอัตราการเติบโต (Growth rate) ของการใช้ไฟฟ้าแต่ละเขตเก็บไว้ด้วย

2. **CALCULATE AVERAGES FOR QUADRANTS**: คำนวณค่าเฉลี่ยของของปริมาณการใช้ไฟฟ้า และอัตราการเติบโต เพื่อใช้สำหรับจัดกลุ่มกาใช้ไฟฟ้าของแต่ละเขตคร่าวๆ เช่น กลุ่มที่ใช้ไฟฟ้าสูงและมีอัตราการเติบโตสูง(กว่าค่าเฉลี่ย)

3. **PLOT BUBBLE CHART (Quadrant Analysis)**: สร้างกราฟข้อมูล โดยขนาดของ ฺBubble แปรผันตามปริมาณการใช้ไฟฟ้าของเขตนั้น

In [27]:
# 1. PREPARE DATA
# Filter for 2024 and 2025
df_2025 = df.loc['2025-01-01':'2025-11-30', office_names].sum()
df_2024 = df.loc['2024-01-01':'2024-11-30', office_names].sum()

# Create a ranking dataframe
ranking_df = pd.DataFrame({
    'Area': df_2025.index,
    'Total_Consumption_2025': df_2025.values,
    'Growth_Rate': ((df_2025 - df_2024) / df_2024) * 100
})

# 2. CALCULATE AVERAGES FOR QUADRANTS
avg_consumption = ranking_df['Total_Consumption_2025'].mean()
avg_growth = ranking_df['Growth_Rate'].mean()

# 3. PLOT BUBBLE CHART (Quadrant Analysis)
fig = px.scatter(
    ranking_df,
    x='Total_Consumption_2025',
    y='Growth_Rate',
    text='Area',
    size='Total_Consumption_2025', # Larger bubble = more consumption
    color='Growth_Rate',
    color_continuous_scale='RdYlGn', # Red (Low) to Green (High) growth
    labels={'Total_Consumption_2025': 'Total Consumption (2025)', 'Growth_Rate': 'YoY Growth Rate (%)'},
    title='<b>Electricity Consumption:</b> District Ranking & Growth 2024 VS 2025'
)

# Add Quadrant Lines
fig.add_vline(x=avg_consumption, line_dash="dash", line_color="gray", opacity=0.5)
fig.add_hline(y=avg_growth, line_dash="dash", line_color="gray", opacity=0.5)

# Annotate Quadrants
fig.add_annotation(x=ranking_df['Total_Consumption_2025'].max(), y=ranking_df['Growth_Rate'].max(),
                   text="<b>Top Right:</b> High Usage, High Growth", showarrow=False, ay=-20)
fig.add_annotation(x=ranking_df['Total_Consumption_2025'].min(), y=ranking_df['Growth_Rate'].max(),
                   text="<b>Top Left:</b> Low Usage, High Growth", showarrow=False, ay=-20)

fig.update_traces(textposition='top center')
fig.update_layout(
    template='plotly_dark',
    height=700,
    xaxis_title="Total Electricity Consumption (2025)",
    yaxis_title="Growth Rate (%) vs 2024"
)

fig.show()

* จากกราฟจะเห็นทุกเขตมี Year-over-Year (YoY) growth rate หรืออัตราการเติบโตของการใช้ไฟฟ้าของแต่ละเขตในปี 2025 ติดลบ ซึ่งตรงกับข้อมูลโดยรวมที่แสดงในกราฟข้อที่ 1 ว่าในปี 2025 นี้ มีแนวโน้มการใช้ไฟฟ้าลดลงจากปี 2024

* เมื่อพิจารณาแต่ละเขตว่าจะเห็นว่ากลุ่มที่ใช้ไฟฟ้าสูงกว่าค่าเฉลี่ยและยังมีอัตราการเติบโตสูงกว่าค่าเฉลี่ย คือ **เขตบางพลี เขตสมุทรปราการ เขตคลองเ**ตย
อย่างไรก็ตามทั้ง 3 เขตนี้มีอัตราการเติบโตในอัตราติดลบ หรือมีการใช้ไฟฟ้าลดลงจากปี 2024 จึงไม่อาจระบุได้ว่าเป็น เขตที่ใช้ไฟมากและยังโตต่อจากข้อมูลที่พิจารณาของปี 2024-2025

* ส่วนเขตที่มีการใช้ไฟฟ้าต่ำและมีอัตราการเติบโตสูงกว่าค่าเฉลี่ย คือ **เขตบางบัวทอง** ซึ่งอาจระบุได้ว่าเป็นเขตที่ใช้น้อยแต่โตเร็ว และแสดงถึงเขตเมืองกำลังขยายตัว

* เพื่อให้เห็นภาพของอัตราการการเติบโตของการใช้ไฟฟ้าของแต่ละเขตมากขึ้น จึงจะลองพิจารณาข้อมูลในปี 2023-2024 ซึ่งภาพรวมการใช้ไฟฟ้าของทุกเขตในช่วงนี้มีแนวโน้มสูงขึ้น  ขั้นตอนต่อไปจึงจะเป็นการสร้างกราฟที่เหมือนกันสำหรับข้อมูลในช่วงดังกล่าว

In [28]:
# 1. PREPARE DATA
# Filter for 2024 and 2025
df_2023 = df.loc['2023-01-01':'2023-12-31', office_names].sum()
df_2024 = df.loc['2024-01-01':'2024-12-31', office_names].sum()

# Create a ranking dataframe
ranking_df = pd.DataFrame({
    'Area': df_2024.index,
    'Total_Consumption_2024': df_2024.values,
    'Growth_Rate': ((df_2024 - df_2023) / df_2023) * 100
})

# 2. CALCULATE AVERAGES FOR QUADRANTS
avg_consumption = ranking_df['Total_Consumption_2024'].mean()
avg_growth = ranking_df['Growth_Rate'].mean()

# 3. PLOT BUBBLE CHART (Quadrant Analysis)
fig = px.scatter(
    ranking_df,
    x='Total_Consumption_2024',
    y='Growth_Rate',
    text='Area',
    size='Total_Consumption_2024', # Larger bubble = more consumption
    color='Growth_Rate',
    color_continuous_scale='RdYlGn', # Red (Low) to Green (High) growth
    labels={'Total_Consumption_2024': 'Total Consumption (2024)', 'Growth_Rate': 'YoY Growth Rate (%)'},
    title='<b>Electricity Consumption:</b> District Ranking & Growth 2023 VS 2024'
)

# Add Quadrant Lines
fig.add_vline(x=avg_consumption, line_dash="dash", line_color="gray", opacity=0.5)
fig.add_hline(y=avg_growth, line_dash="dash", line_color="gray", opacity=0.5)

# Annotate Quadrants
fig.add_annotation(x=ranking_df['Total_Consumption_2024'].max(), y=ranking_df['Growth_Rate'].max(),
                   text="<b>Top Right:</b> High Usage, High Growth", showarrow=False, ay=-20)
fig.add_annotation(x=ranking_df['Total_Consumption_2024'].min(), y=ranking_df['Growth_Rate'].max(),
                   text="<b>Top Left:</b> Low Usage, High Growth", showarrow=False, ay=-20)

fig.update_traces(textposition='top center')
fig.update_layout(
    template='plotly_dark',
    height=700,
    xaxis_title="Total Electricity Consumption (2024)",
    yaxis_title="Growth Rate (%) vs 2023"
)

fig.show()

จากกราฟของข้อมูลการใช้ไฟฟ้าในแต่ละเขตของปี 2024 และอัตราการเติบโตเมื่อเทียบกับปี 2023 จะเห็นว่าภาพรวมทุกเขตมีอัตราการเติบโตเป็นค่าบวกโดย
* เขตที่มีการใช้ไฟฟ้าสูงและอัตราการเติบโตสูงกว่าค่าเฉลี่ยคือ**เขตคลองเตย**

* ส่วนเขตที่มีการใช้ไฟฟ้าต่ำกว่าค่าเฉลี่ย แต่มีอัตราการเติบโตที่สูงกว่าค่าเฉลี่ย คือ **เขตบางบัวทอง เขตบางใหญ่ เขตธนบุรี และเขตนวลจันทร์** ซึ่งแสดงถึงแสดงถึงเขตเมืองกำลังขยายตัว ในปี 2023 ไป 2024

* อย่างไรก็ตามหากพิจารณาร่วมกับข้อมูลของกราฟก่อนหน้าในปี 2025 ซึ่งเหลือเพียง เขตบางบัวทอง ที่ยังมีอัตราการเติบโตของการใช้ไฟฟ้าสูงกว่าค่าเฉลี่ย

หากต้องสรุปผลจากข้อมูลที่มีอยู่นี้
* **เขตคลองเตย** คือ เขตที่ใช้ไฟมากและยังโตต่อ
* **เขตบางบัวทอง** คือเขตที่ใช้น้อยแต่โตเร็วซึ่งอาจสะท้อน พื้นที่เมืองที่กำลังขยายตัว

### **ข้อ 3 — รูปแบบการใช้ไฟฟ้าตามฤดูกาลต่างกันระหว่างเขตไหม?**

Seasonal Pattern by District

*โจทย์* จากข้อมูลรายเดือนของทุกเขต จงสร้างกราฟที่เปรียบเทียบ **รูปแบบฤดูกาลของการใช้ไฟฟ้าระหว่างเขตต่าง ๆ** และระบุให้ได้ว่า
* เขตใดมี ความผันผวนตามฤดูกาลสูงผิดปกติ
* เขตใดมี การใช้ไฟฟ้าคงที่ตลอดปี

พร้อมตั้งสมมติฐานว่าเกิดจากลักษณะการใช้พลังงานแบบใด เช่น
* เขตอุตสาหกรรม
* เขตพาณิชย์
* เขตที่อยู่อาศัย

---

ในข้อนี้จะทำการ normalize ข้อมูลโดยใช้ min-max scaling เพื่อปรับสเกลข้อมูลของทุกเขตให้มาอยู่ในช่วง 0 ถึง 1 เท่ากัน วิธีนี้จะทำให้เราเห็นรูปทรงของการใช้ไฟชัดเจน มีขั้นตอนดังนี้
1. **Prepare the dataframe** : extract seasonal shape ข้อมูลของแต่ละเขต (คล้ายกับวิธีในข้อ 1) ข้อมูลที่ได้จะมี period=12 คือวนซ้ำ/เหมือนกันทุกปี จึงดึงข้อมูลออกมา 12 แถว (แถวที่ 13 จะเท่ากับแถว 1) นำมาเก็บค่า ใน dataframe ที่เตรียมไว้ พร้อมทั้งคำนวณค่าสูงสุดและต่ำสุดของเขตนั้น และทำการ scaling
2. **Identify fluctuation**: หาค่าส่วนเบี่ยงเบนมาตรฐาน std ของข้อมูลแต่ละเขตจากข้อมูล seasonal shape เปรียบเทียบกันแต่ละเดือน (เขตไหนมีการค่าผันผวนแต่ละเดือนสูงจะได้ค่า std สูง, คงที่ตลอดปี std ต่ำ)
3. **Visualization**: สร้างกราฟ โดยเขตที่มีค่า fluctuation สูงและต่ำที่สุดจะ highlight ให้เห็นชัดเจน พร้อมระบุด้านล่างกราฟ

เนื่องจากในโจทย์ข้อนี้จะต้องสรุปตอนท้ายว่าเขตใดมีลักษณะการใช้พลังงานแบบใด ซึ่งจะใช้สมมติฐานดังนี้
* **เขตอุตสาหกรรม**: เขตที่มีค่า Standard Deviation ต่ำ, การใช้พลังงานค่อนข้างคงที่ตลอดปี (Flat Line)
* **เขตพาณิชย์**: มีความผันผวนปานกลาง + Peak หลายช่วง อาจมีจุดสูงสุดในเดือนที่มีเทศกาลหรือวันหยุด
* **เขตที่อยู่อาศัย**: มีความผันผวนสูง (High Variance) มี Peak ชัดเจนในหน้าร้อน (มี.ค. - พ.ค.) พฤติกรรมขึ้นอยู่กับสภาพอากาศสูงมาก (Air Conditioning Load) เมื่ออุณหภูมิสูงขึ้น การใช้ไฟจะพุ่งสูงทันทีและลดลงในหน้าหนาว

In [29]:
# 1. Prepare a dataframe to hold the seasonal shapes
# Index 1-12 represents Jan to Dec
seasonal_shapes_all = pd.DataFrame(index=range(1, 13))

for area in office_names:
    # Perform standard decomposition (Monthly period=12)
    # Use extrapolate_trend to handle potential NaNs at the edges
    decomp = seasonal_decompose(df[area], model='additive', period=12, extrapolate_trend='freq')

    # Extract one full cycle (12 months) of the seasonal component
    # Take the first 12 months as the representative "pattern"
    pattern = decomp.seasonal.iloc[:12].values

    # Min-Max Scaling: Scale the pattern between 0 and 1
    # 1.0 = The month with the highest typical usage in that area
    # 0.0 = The month with the lowest typical usage in that area
    s_min, s_max = pattern.min(), pattern.max()
    seasonal_shapes_all[area] = (pattern - s_min) / (s_max - s_min)

# 2. Identify fluctuation (Standard Deviation of the Seasonal Shape)
# High Std Dev = Highly seasonal (Peaks/Valleys)
# Low Std Dev = Flat/Stable throughout the year
fluctuation = seasonal_shapes_all.std().sort_values(ascending=False)

# 3. Visualization
fig = go.Figure()
months = df['Month'].unique()

for area in office_names:
    # Increase line width for the most and least volatile for visibility
    width = 4 if (area == fluctuation.index[0] or area == fluctuation.index[-1]) else 1.5
    opacity = 1.0 if width == 4 else 0.4

    fig.add_trace(go.Scatter(
        x=months,
        y=seasonal_shapes_all[area],
        name=area,
        line=dict(width=width),
        opacity=opacity,
        hovertemplate='<b>' + area + '</b><br>Month: %{x}<br>Relative Load: %{y:.2f}'
    ))

fig.update_layout(
    title='<b>Seasonal Pattern by District (2018-2025)</b>',
    xaxis_title='Months',
    yaxis_title='Normalized Seasonal Intensity',
    template='plotly_dark',
    hovermode='closest'
)

fig.show()

# Print the Findings
print(f"Highest Seasonal Volatility: {fluctuation.index[0]}")
print(f"Most Stable Consumption: {fluctuation.index[-1]}")

Highest Seasonal Volatility:  คลองเตย 
Most Stable Consumption:  ลาดกระบัง 


จากการคำนวณค่าความผันผวนเพื่อวิเคราะห์การใช้พลังงานไฟฟ้ารายเขตจากข้อมูลปี 2562 - 2568 จะเห็นว่าเขตที่มีการใช้พลังงานคงที่ตลอดทั้งปีมากที่สุดในบรรดาทุกเขตคือ **เขตลาดกระบัง** ซึ่งอาจจะสามารถอนุมานได้ว่าเขตนี้เป็นเขตที่มีลักษณะการใช้พลังงานแบบ**เขตอุตสาหกรรม** ส่วนเขตที่มีความผันผวนของการใช้พลังงานไฟฟ้าสูงที่สุดคือ **เขตคลองเตย** ซึ่งอาจจะสามารถอนุมานได้ว่าเป็นเขตที่มีการใช้พลังงานแบบ**เขตที่อยู่อาศัย**

อย่างไรก็ตาม หากพิจารณาจากกราฟร่วมด้วยและตามบริบทของ**ประเทศไทย** ซึ่งมีช่วงฤดูร้อนที่อุณหภูมิสูงที่สุดในเดือนมีนาคมจนถึงเดือนพฤษภาคม หากคิดตามหลักความต้องการ **การใช้ไฟฟ้าสูงมากในทั้ง 3 เดือนนี้** หากเขตที่พิจารณาเป็นเขตพื้นอยู่อาศัยเป็นหลัก ควรจะสังเกตเห็นความต้องการใช้ไฟฟ้าพุ่งสูงทั้ง 3 เดือนนี้ เมื่อเทียบกับเดือนที่เหลือ แต่เราพบว่า**เขตคลองเตยมีความต้องการการใช้ไฟฟ้าลดลงอย่างมากในเดือนเมษายน**เมื่อเทียบกับเดือนมีนาคม และน้อยกว่าเดือนในฤดูฝน โดยข้อสังเกตนี้ สามารถสันนิษฐานได้ว่ามาจากเหตุผลดังต่อไปนี้
1. ในเดือนเมษายนตรงกับช่วงวันหยุดยาวคือวันสงกรานต์ ทำให้มีการเดินทางกลับต่างจังหวัดมากขึ้น แต่เราควรเห็นผลเดียวกันนี้กับเขตอื่นๆเช่นกัน หากเราอนุมานว่าคนต่างจังหวัดที่เข้ามาอยู่อาศัยมีการกระจายตัวในแต่ละเขตใกล้เคียงกัน
2. หากพิจารณาผลของ covid-19 ซึ่งเป็นค่าผิดปกตินั้นมีผลกับเขตคลองเตยมากที่สุด จากข้อสังเกตในหัวข้อ 3.2 ดังนั้นในการพิจารณาเพื่อหา seasonal pattern ต่อจากนี้ จะ**ใช้ข้อมูลหลัง covid คือปี 2021-2025**
3. พิจารณาจากเหตุผลในข้อ 1 หากวันหยุดยาวมีผลต่อการคิดค่าความผันผวนซึ่งมีผลต่อการจำแนกรูปแบบการใช้ไฟฟ้า เช่น บางเขตอาจเป็นพื้นที่โรงงานอุตสาหกรรม แต่โรงงานหยุดเนื่องจากวันหยุดยาวในเดือนธันวาคม-มกราคม และเดือนเมษายน หากเราลองพิจารณาค่าความผันผวนโดยตัดข้อมูลของทั้ง 3 เดือนนี้ไปจะให้ผลอย่างไร


---

ดังนั้นการพิจารณาจากนี้จะพิจารณาทั้งรูปร่างของกราฟ/รูปแบบการใช้ไฟฟ้า ในแต่ละเดือน ร่วมกับค่าความผันผวน เพื่อจัดกลุ่มและระบุว่าเป็นเขตที่มีการใช้ไฟฟ้าแบบใด โดยในขั้นตอนต่อไป เราจะใช้ข้อมูลหลัง covid-19 คือปี 2021-2025 มาพิจารณาและสร้างกราฟโดยใช้ขั้นตอนเดียวกับก่อนหน้านี้

In [30]:
# 1. Prepare a dataframe to hold the seasonal shapes
# Index 1-12 represents Jan to Dec
seasonal_shapes = pd.DataFrame(index=range(1, 13))

for area in office_names:
    # Perform standard decomposition (Monthly period=12)
    # Use extrapolate_trend to handle potential NaNs at the edges
    decomp = seasonal_decompose(df[area].loc['2021-01-01':'2025-11-30'], model='additive', period=12, extrapolate_trend='freq')

    # Extract one full cycle (12 months) of the seasonal component
    # Take the first 12 months as the representative "pattern"
    pattern = decomp.seasonal.iloc[:12].values

    # Min-Max Scaling: Scale the pattern between 0 and 1
    # 1.0 = The month with the highest typical usage in that area
    # 0.0 = The month with the lowest typical usage in that area
    s_min, s_max = pattern.min(), pattern.max()
    seasonal_shapes[area] = (pattern - s_min) / (s_max - s_min)

# 2. Identify fluctuation (Standard Deviation of the Seasonal Shape)
# High Std Dev = Highly seasonal (Peaks/Valleys)
# Low Std Dev = Flat/Stable throughout the year
fluctuation = seasonal_shapes.std().sort_values(ascending=False)

# 3. Visualization
fig = go.Figure()
months = df['Month'].unique()

for area in office_names:
    # Increase line width for the most and least volatile for visibility
    width = 4 if (area == fluctuation.index[0] or area == fluctuation.index[-1]) else 1.5
    opacity = 1.0 if width == 4 else 0.4

    fig.add_trace(go.Scatter(
        x=months,
        y=seasonal_shapes[area],
        name=area,
        line=dict(width=width),
        opacity=opacity,
        hovertemplate='<b>' + area + '</b><br>Month: %{x}<br>Relative Load: %{y:.2f}'
    ))

fig.update_layout(
    title='<b>Seasonal Pattern by District (2021-2025)</b>',
    xaxis_title='Months',
    yaxis_title='Normalized Seasonal Intensity',
    template='plotly_dark',
    hovermode='closest'
)

fig.show()

# Print the Findings
print(f"Highest Seasonal Fluctuation: {fluctuation.index[0]}")
print(f"Most Stable Consumption: {fluctuation.index[-1]}")

Highest Seasonal Fluctuation:  สามเสน 
Most Stable Consumption:  ราษฎร์บูรณะ 


จากการคำนวณค่าความผันผวนเพื่อวิเคราะห์การใช้พลังงานไฟฟ้ารายเขตจากข้อมูลปี 2564 - 2568 จะเห็นว่าเขตที่มีการใช้พลังงานคงที่ตลอดทั้งปีมากที่สุดในบรรดาทุกเขตคือ **เขตราษฎร์บูรณะ**  ส่วนเขตที่มีความผันผวนของการใช้พลังงานไฟฟ้าสูงที่สุดคือ **เขตสามเสน**

เมื่อพิจารณากราฟที่ได้ จะเห็นว่าถึงแม้ความผันผวนของการใช้ไฟฟ้าเขตราษฎร์บูรณะจะมีค่าต่ำที่สุดในทุกเขต สามารถอนุมานได้ว่าเป็นเขตที่มีการใช้ไฟฟ้าแบบเขตอุตสาหกรรม ซึ่งมีการใช้ไฟฟ้าคงที่ตลอดทั้งปี เราก็ยังสังเกตเห็นปริมาณการใช้ไฟฟ้าลดลงเป็นอย่างมากในเดือนเมษายน ซึ่งสาเหตุอาจจะตรงกับข้อสันนิษฐานเกี่ยวกับ**วันหยุดยาวของโรงงานอุตสาหกรรมในเดือนเมษายน รวมถึงเดือนธันวาคม-มกราคม** ดังนั้นในขั้นตอนต่อไปจะคิดค่าความผันผวนโดย**ตัดเดือนที่มีวันหยุดยาวออกไป**และจัดอันอันดับ เพื่อหาเขตอื่นๆที่มีการใช้ไฟฟ้าแบบเขตอุตสาหกรรมและจัดกลุ่มต่อไป

In [31]:
# Calculate fluctuation from the data where Jan, April, Dec are droped
fluctuation_noholidays = seasonal_shapes.drop([1,4,12]).std().sort_values(ascending=False)
fig = px.bar(fluctuation_noholidays,
             x=fluctuation_noholidays.index,
             y=fluctuation_noholidays.values)

# Plot the bar graph
fig.update_layout(
    title='<b>Fluctuation without Jan, April, Dec by Districts</b>',
    xaxis_title='Areas',
    yaxis_title='Fluctuation',
    template='plotly_dark',
    hovermode='closest'
)
fig.show()

จากกราฟจะเห็นว่าเขตที่มีความผันผวนของการใช้ไฟฟ้าตลอดทั้งปีน้อยที่สุด 4 อันดับคือ ลาดกระบัง สมุทรปราการ บางพลี และ ราษฎร์บูรณะ และเขตที่มีความผันผวนมากที่สุด 4 อันดับคือ คลองเตย สามเสน วัดเลียบ และบางกะปิ แต่จะสังเกตได้ว่า ค่าคววามผันผันของแต่ละเขตไม่แตกต่างกันมากนัก และไม่สามารถจัดกลุ่มของข้อมูลโยใช้ค่้าความผันผวนเพียงอย่างเดียวได้

โดยในการจัดกลุ่มต่อไปนี้จะใช้ทั้งรูปร่างของกราฟ seasonal pattern, ค่าความผันผวน รวมถึงพิจารณาปริมาณการใช้ไฟฟ้ารวมจากข้อที่ 2 ร่วมด้วย

In [32]:
# 1. Define the grops
group1 = [' สมุทรปราการ ', ' บางพลี ', ' ราษฎร์บูรณะ ']
group2 = [' ลาดกระบัง ']
group3 = [' คลองเตย ',  ' บางขุนเทียน ']
group4 = [' วัดเลียบ ',' ยานนาวา ',' บางกะปิ ',' มีนบุรี ',' สามเสน ',
          ' นนทบุรี ',' บางใหญ่ ',' ธนบุรี ',' บางเขน ',' บางบัวทอง ',
          ' นวลจันทร์ ',' บางนา ']
all_groups = [group1, group2, group3, group4]

# 2. Grab the standard Plotly color palette
colors = px.colors.qualitative.Plotly

# 3. Create the 2x1 Subplot Grid
fig = make_subplots(
    rows=2, cols=2,
    subplot_titles=("Group 1: Industrial Sector", "Group 2: Industrial or <br> Mixed-Use, more industrial", "Group 3: Mixed-Use: more residential", "Group 4: Residential Sector"),
    shared_yaxes=True,     # Consistent 0-1 scale for all plots
    vertical_spacing=0.15, # Space between top and bottom rows
    horizontal_spacing=0.1 # Space between left and right columns
)

# 4. Add traces to subplots
# Using a counter to cycle through colors across all 18 lines
color_idx = 0

for i, current_group in enumerate(all_groups):
    # Calculate grid position (row 1-2, col 1-2)
    row = (i // 2) + 1
    col = (i % 2) + 1

    for area in current_group:
        if area in seasonal_shapes.columns:
            fig.add_trace(
                go.Scatter(
                    x=months,
                    y=seasonal_shapes[area],
                    name=area,
                    line=dict(width=2, color=colors[color_idx % len(colors)]),
                    mode='lines',
                    hovertemplate='<b>%{fullData.name}</b><br>Month: %{x}<br>Value: %{y:.2f}'
                ),
                row=row, col=col
            )
            color_idx += 1

# 5. Global Layout Styling
fig.update_layout(
    title='<b>Seasonal Patterns by Disttricts: </b> four groups',
    template='plotly_dark',
    height=800,
    legend=dict(
        orientation="h",
        yanchor="top",
        y=-0.15,
        xanchor="center",
        x=0.5
    ),
    margin=dict(t=100, b=150, l=50, r=50)
)

# Set axis titles
fig.update_xaxes(title_text="Months")
fig.update_yaxes(title_text="Normalized Intensity", col=1)

fig.show()

จากข้อมูลข้างต้นทั้งหมด จะสามารถแบ่งกลุ่มของเขตตามรูปแบบการใช้ไฟฟ้าได้ทั้งหมด 4 กลุ่ม
1. **กลุ่มที่ 1 เขตอุตสาหกรรม** ได้แก่ **เขตสมุทรปราการ เขตบางพลี และเขตราษฎร์บูรณะ** ซึ่งมีการใช้ไฟฟ้าคงที่ตลอดทั้งปี (หากคำนวณแบบไม่รวมวันหยุดยาว) สังเกตเห็นค่าการใช้ไฟฟ้าที่ลดลงอย่างมากในเดือนเมษายน ซึ่งอาจเป็นผลมาจากวันหยุดยาวของโรงงานอุตสาหกรรม และจะมีปริมาณการใช้ไฟฟ้าสูงกว่าค่าเฉลี่ย (จากกราฟในข้อ 2)
2. **กลุ่มที่ 2 เขตการใช้ไฟฟ้าแบบผสม โดยมีเขตอุตสาหกรรมเป็นส่วนใหญ่** หรืออาจเป็น**เขตอุตสาหกรรม** ได้แก่ **เขตลาดกระบัง** ซึ่างมีการใช้ไฟฟ้าคงที่ตลอดทั้งปีทั้งการคำนวณแบบรวมวันหยุดยาวและไม่รวมวันหยุดยาว อย่างไรก็ตาม เนื่องจากมีปริมาณการใช้ไฟฟ้ารวมที่ต่ำกว่าค่าเฉลี่ย และสังเกตเห็นว่าในเดือนเมษายนซึ่งเป็นวันหยุดยาว ปริมาณการใช้ไฟฟ้าไม่ได้ลดลงแบบกลุ่มที่ 1 จึงอาจเป็นไปได้ว่า อาจมีการใช้พื้นที่แบบเขตอยู่อาศัยหรือเขตพาณิชย์รวมอยู่ด้วย แต่พื้นที่ส่วนใหญ่ยังเป็นเขตอุตสาหกรรม(ทั้งนี้อาจตีความว่าวันหยุดยาวไม่ได้มีผลกระทบต่อปริมาณการใช้ไฟฟ้าแบบกลุ่มที่ 1 จึงยังจัดเป็นเขตอุตสาหกรรมก็ได้)
3. **กลุ่มที่ 3 เขตการใช้ไฟฟ้าแบบผสม โดยมีเขตที่อยู่อาศัยเป็นส่วนใหญ่** ได้แก่ **เขตคลองเตยและเขตบางขุนเทียน** ซึ่งมีความผันผวนของปริมาณการใช้ไฟฟ้ามาก แต่ยังสังเกตเห็นว่าในเดือนเมษายนมีปริมาณการใช้ไฟฟ้าลดลงจากเดือนมีนาคม อาจเป็นไปได้ว่ามีผลของการใช้ไฟฟ้าแบบเขตอุตสาหกรรม แบบกลุ่มที่ 1 (วันสงกรานต์มีการหยุดการทำงานของโรงงานอุตสาหกรรม)

4. **กลุ่มที่ 4 เขตที่อยู่อาศัย** ได้แก่ เขตที่เหลือทั้งหมด คือ วัดเลียบ  ยานนาวา บางกะปิ มีนบุรี สามเสน นนทบุรี  บางใหญ่  ธนบุรี  บางเขน บางบัวทอง นวลจันทร์ และบางนา ซึ่งมีการใช้ไฟฟ้าผันผวนตลอดปี และเพิ่มขึ้นในฤดูร้อนทั้งหมด

ทั้งนี้ ยังไม่สามารถแยกเขตพาณิชย์ออกมาจากข้อมูลที่มีได้ ด้วยเทคนิคที่ใช้อยู่ เนื่องจากค่าความผันผวนที่คำนวณได้มีค่าไม่ต่างกันมากนัก จึงยังไม่สามารถจัดกลุ่มเขตพาณิชย์ ที่มีค่าความผันผวนปานกลางออกมาได้ หากมีข้อมูลเพิ่มเติม เช่น ลักษณะการใช้ไฟฟ้าของแต่ละเขตใน 1 วัน อาจจะสามารถช่วยให้แยกแยะได้ เนื่องจากเขตพาณิชย์อาจมีการใช้ไฟฟ้าสูงในตอนกลางวันและน้อยลงในตอนกลางคืนจากอาคารสำนักงานและห้างสรรรพสินค้า ซึ่งต่างจากเขตอุตสาหกรรมที่มีการใช้ไฟฟ้าสูงตลอดทั้งวัน และเขตพื้นที่อาศัยที่อาจมีการใช้ไฟฟ้าสูงขึ้นในตอนกลางคืน